src/aspire/image/__init__.py

```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
+            Alternatively, it can be a column vector of length 2, in which case the same shifts is applied to each image.
+        :return: The Image translated by the shifts, with periodic boundaries.
+        """
+        if shifts.ndim == 1:
+            shifts = shifts[np.newaxis, :]
+
+        warnings.warn('Multiple implementations of im_translate. Running both for sanity check.'
+                      'Eliminating this will result in substantial speedup.')
+
+        im_translated = _im_translate(self.data, shifts)       # slower
+        im_translated2 = _im_translate2(self.data, -shifts.T)  # faster
+        assert np.allclose(im_translated, im_translated2)
+
+        return Image(im_translated)
+
+    def downsample(self, res):
+        """
+        Downsample Image to a specific resolution. This method returns a new Image.
+        :param res: int - new resolution, should be <= the current resolution of this Image
+        :return: The down-sampled Image object.
+        """
+        grid = grid_2d(self.res)
+        grid_ds = grid_2d(res)
+
+        im_ds = np.zeros((res, res, self.n_images)).astype(self.data.dtype)
+
+        # x, y values corresponding to 'grid'. This is what scipy interpolator needs to function.
+        res_by_2 = self.res / 2
+        x = y = np.ceil(np.arange(-res_by_2, res_by_2)) / res_by_2
+
+        mask = (np.abs(grid['x']) < res / self.res) & (np.abs(grid['y']) < res / self.res)
+        im = np.real(centered_ifft2(centered_fft2(self.data) * np.expand_dims(mask, 2)))
+
+        for s in range(im_ds.shape[-1]):
+            interpolator = RegularGridInterpolator(
+                (x, y),
+                im[:, :, s],
+                bounds_error=False,
+                fill_value=0
+            )
+            im_ds[:, :, s] = interpolator(np.dstack([grid_ds['x'], grid_ds['y']]))
+
+        return Image(im_ds)
+
+    def filter(self, filter=None, filter_values=None):
+        """
+        Apply a Filter object to the Image. This method returns a new Image.
+        :param filter: An object of type Filter
+        :param filter_values: An ndarray of precomputed filter values that can be specified instead of a Filter object.
+        :return: A new filtered Image object.
+        """
+        ensure((filter is None and filter_values is not None) or (filter is not None and filter_values is None),
+               'Specify either the Filter object or pre-evaluated filter values directly.')
+
+        if filter_values is None:
+            filter_values = filter.evaluate_grid(self.res)
+
+        im_f = centered_fft2(self.data)
+        if im_f.ndim > filter_values.ndim:
+            im_f = np.expand_dims(filter_values, 2) * im_f
+        else:
+            im_f = filter_values * im_f
+        im = centered_ifft2(im_f)
+        im = np.real(im)
+
+        return Image(im)
 
     def rotate(self):
         raise NotImplementedError
 
+    def save(self, mrcs_filepath, overwrite=False):
```

Comment:
Probably a generalized output function is enough. In the derived class it can be overrided in order to save to mrc files. I think that capability of outputting to new mrc files is import to save the processed images such as denoising or downsampling.

Reply:
Opened a new issue #48 for this. Closing here.